cmd/metal-api/internal/datastore/image.go

```diff
@@ -35,3 +47,146 @@ func (rs *RethinkStore) DeleteImage(i *metal.Image) error {
+		// Always keep the most recent image for one OS even if no machine uses it is not valid anymore
+		// this prevents that there is no image at all left if no new images are pushed.
+		_, ok := firstOSImage[image.OS]
+		if !ok {
+			firstOSImage[image.OS] = true
+			continue
+		}
+
+		if isOrphanImage(image, machines) {
+			err := rs.DeleteImage(&image)
+			if err != nil {
+				return nil, fmt.Errorf("unable to delete image:%s err:%v", image.ID, err)
+			}
+			result = append(result, image)
+		}
+	}
+	return result, nil
+}
+
+// isOrphanImage check if a image is not allocated and older than allowed.
+func isOrphanImage(image metal.Image, machines *metal.Machines) bool {
+	if time.Since(image.ExpirationDate) < 0 {
+		return false
+	}
+	orphan := true
+	for _, m := range *machines {
+		if m.Allocation == nil {
+			continue
+		}
+		if image.ID == m.Allocation.ImageID {
+			orphan = false
+			continue
+		}
+	}
+	return orphan
+}
+
+// getMostRecentImageFor
+// the id is in the form of: <name>-<version>
+// where name is for example ubuntu or firewall
+// version must be a semantic version, see https://semver.org/
+// we decided to specify the version in the form of major.minor.patch,
+// where patch is in the form of YYYYMMDD
+// If version is not fully specified, e.g. ubuntu-19.10 or ubuntu-19.10
+// then the most recent ubuntu image (ubuntu-19.10.20200407) is returned
+// If patch is specified e.g. ubuntu-20.04.20200502 then this exact image is searched.
+func (rs *RethinkStore) getMostRecentImageFor(id string, images metal.Images) (*metal.Image, error) {
+	os, sv, err := rs.GetOsAndSemver(id)
+	if err != nil {
+		return nil, err
+	}
+
+	matcher := "~"
+	// if patch is given return a exact match
+	if sv.Patch() > 0 {
+		matcher = "="
+	}
+	constraint, err := semver.NewConstraint(matcher + sv.String())
+	if err != nil {
+		return nil, fmt.Errorf("could not create constraint of image version:%s err:%v", sv, err)
+	}
+
+	var latestImage *metal.Image
+	sortedImages := sortImages(images)
+	for _, image := range sortedImages {
+		if !strings.HasPrefix(id, image.OS) {
+			continue
+		}
+		if time.Since(image.ExpirationDate) > 0 {
+			continue
+		}
+		v, err := semver.NewVersion(image.Version)
+		if err != nil {
+			continue
+		}
+		if constraint.Check(v) {
+			latestImage = &image
+			break
+		}
+	}
+	if latestImage != nil {
+		return latestImage, nil
+	}
+	return nil, fmt.Errorf("no image for os:%s version:%s found", os, sv)
+}
+
+// GetOsAndSemver parses a imageID to OS and Semver, or returns an error
+func (rs *RethinkStore) GetOsAndSemver(id string) (string, *semver.Version, error) {
+	imageParts := strings.Split(id, "-")
+	if len(imageParts) < 2 {
+		return "", nil, fmt.Errorf("image does not contain a version")
+	}
+
+	os := imageParts[0]
+	version := strings.Join(imageParts[1:], "")
+	v, err := semver.NewVersion(version)
+	if err != nil {
+		return "", nil, err
+	}
+	return os, v, nil
+}
+
+func sortImages(images []metal.Image) []metal.Image {
```

Comment:
i don't get if this is a correct sorter for different images. you sort the slice twice, first by the OS, than again by the version.
when you have a list of images:
~~~
ubuntu-20.04.20200401
ubuntu-20.04.20200402
alpine-20.04.20200401
alpine-20.04.20200402
~~~
then after the first sort you have:
~~~
alpine-20.04.20200401
alpine-20.04.20200402
ubuntu-20.04.20200401
ubuntu-20.04.20200402
~~~
and after the seconds sort:
~~~
alpine-20.04.20200401
ubuntu-20.04.20200401
alpine-20.04.20200402
ubuntu-20.04.20200402
~~~
???

i might be wrong ... but what's the first sort for if your second resorts and orders by version?
EDIT: ok i see, as you use the `Stable` sort, you maintain the first sorting order for equal OS'es. couldn't this be done with one single compare-function which does the semver-compare and if both are equal, it returns the os-comparison?

Reply:
Yes, there was even a BUG here, modified implementation and tests to cover this and have a more clear implementation.